### PR TITLE
Validator over Rest

### LINKED ISSUE #279: Implement REST api client in validator
- create REST API client to consume #278 - should implement client interface (might need adjustment)
- adjust validator to choose which api client to use (rpc, rest)

### PR DESCRIPTION
**Pull request by twoeths**
## Goal
+ Close #279

### CODE REVIEW

File: packages/lodestar/src/validator/rest/beacon/beacon.ts [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+  public async getGenesisTime(): Promise<number64> {
+    return this.client.get<number64>("/genesis_time");
+  }
+
+  public async getSyncingStatus(): Promise<boolean | SyncingStatus> {
+    return this.client.get<boolean | SyncingStatus>("/syncing")
+  }
+
+  public async getChainHead(): Promise<BeaconBlock> {
+    // not used in validator and not part of Rest API spec, remove this from IValidatorApi?
```
Comment (mpetrunic) [inline packages/lodestar/src/validator/rest/beacon/beacon.ts]: yes

File: packages/lodestar/src/validator/rest/beacon/beacon.ts [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+    return Promise.reject();
+  }
+
+  public async getBeaconState(): Promise<BeaconState> {
+    // not used in validator and not part of Rest API spec, remove this from IValidatorApi?
```
Comment (mpetrunic) [inline packages/lodestar/src/validator/rest/beacon/beacon.ts]: yes

File: packages/lodestar/src/validator/rest/validator/validator.ts [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+  }
+
+  public async produceBlock(slot: Slot, randaoReveal: bytes96): Promise<BeaconBlock> {
+    const url = `/block?slot=${slot}&randao_reveal=${randaoReveal.toString("hex")}`;
+    return this.client.get<BeaconBlock>(url);
+  }
+
+  public async produceAttestation(validatorPubKey: BLSPubkey, pocBit: boolean, slot: Slot, shard: Shard): Promise<Attestation> {
+    // validator_pubkey and poc_bit are not needed at this time
+    const url = `/attestation?slot=${slot}&shard=${shard}`;
```
Comment (mpetrunic) [inline packages/lodestar/src/validator/rest/validator/validator.ts]: That changed, current merged api includes those

I think

File: packages/lodestar/src/validator/validator.ts [hunk 4]
```diff
@@ -60,6 +61,8 @@ class Validator {
       this.rpcClient = this.opts.rpcInstance;
     } else if(this.opts.rpc) {
       this.rpcClient = new RpcClientOverWs({rpcUrl: this.opts.rpc}, {config: this.config});
+    } else if(this.opts.restUrl) {
```
Comment (mpetrunic) [inline packages/lodestar/src/validator/validator.ts:68]: I would move api client initialization to separate method and probably rename it to api or apiClient or something more generic than rpcClient

**Review by mpetrunic — CHANGES_REQUESTED** [review]
Yeah, that's right direction.

**Review by wemeetagain — APPROVED** [review]
:100: this looks great

**Review by mpetrunic — APPROVED** [review]

**Pull request merged**